test/unit/test_HNF.cc

```diff
@@ -0,0 +1,327 @@
+
+int u[5][5] = {{25, -300, 1050, -1400, 630},
+               {-300, 4800, -18900, 26880, -12600},
+               {1050, -18900, 79380, -117600, 56700},
+               {-1400, 26880, -117600, 179200, -88200},
+               {630, -12600, 56700, -88200, 44100}};
+
+
+for (int i = 0; i < rows; i++)
+for (int j = 0; j < cols; j++)
+ms.MM(U, i+1, j+1) = u[i][j];
+
+//HNF
+
+int hnf_ref[5][5] = {{5, 0, 0, 0, 0},
+                     {0, 60, 0, 0, 0},
+                     {-210, 0, 420, 0, 0},
+                     {-280, 0, 0, 840, 0},
+                     {630, 0, 0, 0, 2520}};
+
+MId H = ms.getNewMatrix(rows, cols);
+
+int dim;
+MId U1 = MId_Undef;
+MId V1 = MId_Undef;
+ms.compute_hnf_v1(U, H, dim, U1, V1);
+cout << "The hnf matrix:" << endl << ms.print(H) << endl;
+for (int i = 0; i < rows; i++)
+for (int j = 0; j < cols; j++)
+ASSERT_EQ(ms.MM(H, i+1, j+1), hnf_ref[i][j]);
+}
+*/
+
+TEST_F(HNF_test, test_hnf8) {
+    int rows = 4;
+    int cols = 5;
+    MId U = ms.getNewMatrix(rows, cols);
+
+    int u[4][5] = {{9,  6,  0, -8,  0},
+                   {-5, -8,  0,  0,  0},
+                   {0,  0,  0,  4,  0},
+                   {0,  0,  0, -5,  0}};
+
+
+    for (int i = 0; i < rows; i++)
+        for (int j = 0; j < cols; j++)
+            ms.MM(U, i+1, j+1) = u[i][j];
+
+    //HNF
+
+    int hnf_ref[4][5] = {{1, 0, 0, 0, 0},
+                         {1, 2, 0, 0, 0},
+                         {28, 36, 84, 0, 0},
+                         {-35, -45, -105, 0, 0}};
+
+    MId H = ms.getNewMatrix(rows, cols);
+
+    int dim;
+    MId U1 = MId_Undef;
+    MId V1 = MId_Undef;
+    ms.compute_hnf_v1(U, H, dim, U1, V1);
+    std::cout << "The hnf matrix:" << std::endl << ms.print(H) << std::endl;
+    for (int i = 0; i < rows; i++)
+        for (int j = 0; j < cols; j++)
+            ASSERT_EQ(ms.MM(H, i+1, j+1), hnf_ref[i][j]);
+}
+
+TEST_F(HNF_test, test_hnf9) {
+    int rows = 6;
+    int cols = 4;
+    MId U = ms.getNewMatrix(rows, cols);
+
+    int u[6][4] = {{-5, -2,  7,  1},
+                   {8,  8,  -5,  -1},
+                   {-3, -2, -8, 6},
+                   {-9, -2, 4, 0},
+                   {5, 8, 3, 8},
+                   {5, 5, -4, -3}};
+
+
+    for (int i = 0; i < rows; i++)
+        for (int j = 0; j < cols; j++)
+            ms.MM(U, i+1, j+1) = u[i][j];
+
+    //HNF
+
+    int hnf_ref[6][4] = {{1, 0, 0, 0},
+                         {0, 1, 0, 0},
+                         {3, 1, 4, 0},
+                         {237, 103, 352, 486},
+                         {-299, -130, -450, -627},
+                         {90, 40, 135, 188}};
+
+    MId H = ms.getNewMatrix(rows, cols);
+
+    int dim;
+    MId U1 = MId_Undef;
+    MId V1 = MId_Undef;
+    ms.compute_hnf_v1(U, H, dim, U1, V1);
+    std::cout << "The hnf matrix:" << std::endl << ms.print(H) << std::endl;
+    for (int i = 0; i < rows; i++)
+        for (int j = 0; j < cols; j++)
+            ASSERT_EQ(ms.MM(H, i+1, j+1), hnf_ref[i][j]);
+}
+
+
```

Comment:
There is something weird with the commit or how GitHub shows me the difference because it shows `test/unit/test_HNF.cc` once as a completly new file and once as with a single line changed.

Reply:
I see it now, the new file is `.cc` file, the old one is `.cpp`.
This can be done with `git mv`, so git automatically picks it up as a renaming operation.

Reply:
It looks like both the `.cc` and `.cpp` exist in the repository, so either I forgot to remove the `.cpp` or I wanted to do this in a different commit.  Anyway, I think I managed to fix this.